language-support/ts/daml-types/index.ts

```diff
@@ -419,4 +420,82 @@ export const TextMap = <T>(t: Serializable<T>): Serializable<TextMap<T>> => ({
   }
 });
 
-// TODO(MH): `Map` type.
+/**
+ * The counterpart of DAML's `DA.Map.Map K V` type.
+ *
+ * @typeparam K The type of the map keys.
+ * @typeparam V The type of the map values.
+ */
+export interface Map<K, V> {
+  get: (k: K) => V | null;
+  getOr: (k: K, ifNotFound: V) => V;
+  has: (k: K) => boolean;
+  insert: (k: K, v: V) => Map<K, V>;
+  remove: (k: K) => Map<K, V>;
+  keys: () => K[];
+  values: () => V[];
+  kvs: () => [K, V][];
+}
+
+// This code assumes that the decoder is only ever used in decoding values
+// straight from the API responses, and said raw responses are never reused
+// afterwards. This should be enforced by this class not being exported and the
+// daml-ledger module not letting raw JSON responses escape without going
+// through this.
+//
+// Without that assumption, the constructor would need to deep-copy its kvs
+// argument.
+class MapImpl<K, V> implements Map<K, V> {
+  private _kvs: [K, V][];
+  private _keys: K[];
+  private _values: V[];
+  constructor(kvs: [K, V][]) {
+    // sorting done so that generic object deep comparison would find equal
+    // maps equal (as defined by jest's expect().toEqual())
+    this._kvs = _.sortBy(kvs, kv => JSON.stringify(kv[0]));
+    this._keys = this._kvs.map(e => e[0]);
+    this._values = this._kvs.map(e => e[1]);
+  }
+  private _idx(k: K): number {
+    return this._keys.findIndex((l) => _.isEqual(k, l));
+  }
+  has(k: K): boolean {
+    return this._idx(k) !== -1;
+  }
+  private _getOr(k: K, o: V | null): V | null {
+    return this.has(k) ? this._values[this._idx(k)] : o;
+  }
+  get(k: K): V | null { return _.cloneDeep(this._getOr(k, null)); }
+  getOr(k: K, ifNotFound: V): V { return _.cloneDeep(this._getOr(k, ifNotFound)) as V; }
+  insert(k: K, v: V): Map<K, V> {
+    if (this.has(k)) {
+      const cpy = this._kvs.slice();
+      cpy[this._idx(k)] = [k, v];
+      return new MapImpl(cpy);
+    } else {
+      const head: [K, V][] = [[k, v]];
+      return new MapImpl(head.concat(this._kvs));
+    }
+  }
+  remove(k: K): Map<K, V> {
+    if (this.has(k)) {
```

Comment:
Same here, you can call `idx_` and work with the return value instead of calling `has` first.

Reply:
I thought the `has` version was more readable, but if you think perf is more important I'll change it.

Reply:
Fine to keep `has`. I guess if we care about performance, we should rather revisit the whole implementation than microoptimizing a few lines here.